### ISSUE TITLE
Protocol.Type.self fail to cast to Protocol 

### ISSUE THREAD

**Issue by miku1958** [apple/swift#62134]
**Describe the bug**
I have a method for manipulating Protocol that looks like this
```
func classConforming<P>(to protocolName: P.Type) -> P? {
	let p: Any = protocolName
	guard let pp = p as? Protocol else {
		return nil
	}
	return self.objonly_classConformingToProtocol(pp) as? P
}

self.objonly_classConformingToProtocol accepts a Protocol object as parameter and return AnyObject
```
When I call it like:
classConforming(to: aProtocol.Type.self)
Xcode shows me that aProtocol.Type.self is aProtocol.Type.Protocol and lldb tells me that protocolName is "@thick aProtocol.Type.Protocol" and p is "" but print will tell me it is aProtocol.Type and I can't cast it to Protocol, pp will always nil.

I also tried adding where P.Type == Protocol, but the compiler would give me an error: 
> Generic signature requires types 'P.Type' and 'Protocol' to be the same

I'm not sure how the compiler converts, but I think since Swift lacks a mechanism to determine whether it's a protocol/class/struct/enum/actor, it should at least ensure type consistency when passing parameters

**Comment by miku1958** [comment]
Another try:
type(of: protocolName) will return me “theProtocol.Type.Protocol”, both lldb and print are, but I still can't cast it to Protocol

**Comment by miku1958** [comment]
One possible solution I expect is to make the type returned by theProtocol.self (theProtocol.Protocol, aka Protocol) consistent with the one in the parameter: 
I can use P.Protocol instead of Protocol, which is still essentially an Objc object, but then I can return P.Type by genericizing it, like this:
```
func classConforming<P>(to protocolName: P.Protocol) -> P.Type? {
	return self.objonly_classConformingToProtocol(protocolName) as? P.Type
}

classConforming(to: aProtocol.self)
```

**Comment by miku1958** [comment]
Hi @AnthonyLatsis,  I know when I use aProtocol.Type.self, the complier will return me a Protocol object, and it will lose all the type information about the aProtocol, and the problem I'm having is that I want to get the Protocol object, and I want the function to be generic, so I can return the object that implements the aProtocol

**Comment by AnthonyLatsis** [comment]
The protocol metatype object (the protocol type as an object) is `MyProtocol.self` (the type of this object is `MyProtocol.Protocol`). `MyProtocol.Type.self` is a higher-level, meta-metatype object; its type is `MyProtocol.Type.Protocol`. Although these meta-metatypes exist as part of the type system, they have little if any practical value. 

Unfortunately, there is no way of expressing a generic parameter that accepts only protocol types, so `func foo<T>(_: T.Protocol)` will not work. 

> `- (Class)objonly_classConformingToProtocol:(Protocol *)inProtocol;`

I wasn’t able to find any documentation on this `Protocol` type. How do you instantiate it? There is no implicit Swift conversion from a protocol metatype like `aProtocol.self` to `Protocol`.

**Comment by miku1958** [comment]
We basically only use the Protocol object as a key to the dictionary to find the class in objonly_classConformingToProtocol, and this practice of using objonly_classConformingToProtocol is very widespread in our projects.

What I mean is, can this generic be added at the compiler level. The compiler can know what I mean when I use P.Protocol like this:
```
func classConforming<P>(to objcProtocol: P.Protocol) -> P.Type? {
	return self.objonly_classConformingToProtocol(objcProtocol) as? P.Type
}
```
Or 

```
func classConforming<P>(to objcProtocol: P.Type) -> P.Type? where P: Protocol {
	return self.objonly_classConformingToProtocol(objcProtocol.Protocol) as? P.Type
}
```

**Comment by AnthonyLatsis** [comment]
> What I mean is, can this generic be added at the compiler level. 

No, you cannot constrain a generic parameter to protocols only, Swift currently lacks this kind of expressivity. There is no analogue of `: AnyObject` for protocols.

> ```swift
> func classConforming<P>(to objcProtocol: P.Type) -> P.Type? where P: Protocol {
>   return self.objonly_classConformingToProtocol(objcProtocol.Protocol) as? P.Type
> }
> ```

The question of whether something similar to this will work boils down to the question of how you instantiate an object of this `Protocol` type in Swift. Have you ever used `objonly_classConformingToProtocol` in Swift successfully before, without generics? If so, could you provide an example?

**Comment by miku1958** [comment]
Hi @AnthonyLatsis, I can use objonly_classConformingToProtocol in Swift directly, but the "Class" from Objc will become "AnyClass" in Swift, and for historical reasons, compiler allow me to call any Objc method with AnyClass. It's not safe, and it's not what I want
